File: frontend/src/api/projects.ts
import axios from 'axios'
import type { Project, ProjectCreate } from '@/types/project'

const api = axios.create({ baseURL: 'http://localhost:8000' })

export const getProjects = (): Promise<Project[]> =>
  api.get<Project[]>('/api/projects').then(r => r.data)

export const getProject = (id: number): Promise<Project> =>
  api.get<Project>(`/api/projects/${id}`).then(r => r.data)

export const createProject = (body: ProjectCreate): Promise<Project> =>
  api.post<Project>('/api/projects', body).then(r => r.data)

export const updateProject = (id: number, body: ProjectCreate): Promise<Project> =>
  api.put<Project>(`/api/projects/${id}`, body).then(r => r.data)

export const deleteProject = (id: number): Promise<void> =>
  api.delete(`/api/projects/${id}`).then(() => undefined)

File: backend/app/routers/projects.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List

from .. import crud, schemas
from ..database import get_db

router = APIRouter(prefix="/api/projects", tags=["projects"])


@router.get("", response_model=List[schemas.Project])
def list_projects(db: Session = Depends(get_db)):
    return crud.get_projects(db)


@router.post("", response_model=schemas.Project, status_code=201)
def create_project(body: schemas.ProjectCreate, db: Session = Depends(get_db)):
    return crud.create_project(db, body)


@router.get("/{id}", response_model=schemas.Project)
def get_project(id: int, db: Session = Depends(get_db)):
    obj = crud.get_project(db, id)
    if not obj:
        raise HTTPException(status_code=404, detail="Project not found")
    return obj


@router.put("/{id}", response_model=schemas.Project)
def update_project(id: int, body: schemas.ProjectCreate, db: Session = Depends(get_db)):
    obj = crud.update_project(db, id, body)
    if not obj:
        raise HTTPException(status_code=404, detail="Project not found")
    return obj


@router.delete("/{id}", status_code=204)
def delete_project(id: int, db: Session = Depends(get_db)):
    if not crud.delete_project(db, id):
        raise HTTPException(status_code=404, detail="Project not found")
